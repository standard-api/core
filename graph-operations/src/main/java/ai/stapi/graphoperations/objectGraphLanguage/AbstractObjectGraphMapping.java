package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.declaration.AbstractDeclaration;

public abstract class AbstractObjectGraphMapping extends AbstractDeclaration
    implements ObjectGraphMapping {

  public static final String DECLARATION_TYPE = "ObjectGraphMapping";
  protected GraphDescription graphDescription;

  protected AbstractObjectGraphMapping() {

  }

  public AbstractObjectGraphMapping(
      GraphDescription graphDescription,
      String deserializationType
  ) {
    super(deserializationType);
    this.graphDescription = graphDescription;
  }

  public AbstractObjectGraphMapping(String deserializationType) {
    super(deserializationType);
  }

  @Override
  public GraphDescription getGraphDescription() {
    return graphDescription;
  }

  @Override
  public String getDeclarationType() {
    return DECLARATION_TYPE;
  }
}
