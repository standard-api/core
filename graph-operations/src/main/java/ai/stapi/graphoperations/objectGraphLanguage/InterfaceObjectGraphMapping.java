package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;

public class InterfaceObjectGraphMapping extends AbstractObjectGraphMapping {

  public static final String SERIALIZATION_TYPE = "InterfaceObjectGraphMapping";
  private String interfaceUuid;

  public InterfaceObjectGraphMapping() {
    super(SERIALIZATION_TYPE);
  }

  public InterfaceObjectGraphMapping(
      String interfaceUuid,
      GraphDescription graphDescription
  ) {
    super(
        graphDescription,
        SERIALIZATION_TYPE
    );
    this.interfaceUuid = interfaceUuid;
  }

  public String getInterfaceUuid() {
    return interfaceUuid;
  }
}
