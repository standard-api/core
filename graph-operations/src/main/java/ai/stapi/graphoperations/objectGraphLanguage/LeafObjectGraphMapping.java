package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;

public class LeafObjectGraphMapping extends AbstractObjectGraphMapping {

  public static final String SERIALIZATION_TYPE = "LeafObjectGraphMapping";

  private LeafObjectGraphMapping() {

  }

  public LeafObjectGraphMapping(GraphDescription graphDescription) {
    super(
        graphDescription,
        SERIALIZATION_TYPE
    );
  }
}
