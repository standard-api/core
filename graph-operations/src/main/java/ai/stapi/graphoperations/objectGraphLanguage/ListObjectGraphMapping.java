package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;

public class ListObjectGraphMapping extends AbstractObjectGraphMapping {

  public static final String SERIALIZATION_TYPE = "ListObjectGraphMapping";
  private ObjectGraphMapping childObjectGraphMapping;

  private ListObjectGraphMapping() {

  }

  public ListObjectGraphMapping(
      GraphDescription graphDescription,
      ObjectGraphMapping childObjectGraphMapping
  ) {
    super(
        graphDescription,
        SERIALIZATION_TYPE
    );
    this.childObjectGraphMapping = childObjectGraphMapping;
  }

  public ObjectGraphMapping getChildObjectGraphMapping() {
    return childObjectGraphMapping;
  }
}
