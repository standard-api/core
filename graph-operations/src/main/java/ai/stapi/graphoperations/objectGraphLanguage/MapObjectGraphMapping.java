package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;

public class MapObjectGraphMapping extends AbstractObjectGraphMapping {

  public static final String SERIALIZATION_TYPE = "MapObjectGraphMapping";

  private ObjectGraphMapping keyObjectGraphMapping;
  private ObjectGraphMapping valueObjectGraphMapping;

  private MapObjectGraphMapping() {

  }

  public MapObjectGraphMapping(
      GraphDescription graphDescription,
      ObjectGraphMapping keyObjectGraphMapping,
      ObjectGraphMapping valueObjectGraphMapping
  ) {
    super(graphDescription, SERIALIZATION_TYPE);
    this.keyObjectGraphMapping = keyObjectGraphMapping;
    this.valueObjectGraphMapping = valueObjectGraphMapping;
  }

  public ObjectGraphMapping getKeyObjectGraphMapping() {
    return keyObjectGraphMapping;
  }

  public ObjectGraphMapping getValueObjectGraphMapping() {
    return valueObjectGraphMapping;
  }
}
