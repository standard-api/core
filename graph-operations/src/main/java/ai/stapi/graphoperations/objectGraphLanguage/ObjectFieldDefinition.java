package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.declaration.Declaration;
import ai.stapi.serialization.AbstractSerializableObject;

public class ObjectFieldDefinition extends AbstractSerializableObject {

  public static final String SERIALIZATION_TYPE = "ObjectFieldDefinition";
  private Declaration relation;
  private ObjectGraphMapping fieldObjectGraphMapping;

  private ObjectFieldDefinition() {
  }

  public ObjectFieldDefinition(
      Declaration relation,
      ObjectGraphMapping fieldObjectGraphMapping
  ) {
    super(SERIALIZATION_TYPE);
    this.fieldObjectGraphMapping = fieldObjectGraphMapping;
    this.relation = relation;
  }

  public ObjectFieldDefinition(
      Declaration relation,
      ObjectGraphMapping fieldObjectGraphMapping,
      String serializationType
  ) {
    super(serializationType);
    this.fieldObjectGraphMapping = fieldObjectGraphMapping;
    this.relation = relation;
  }

  public ObjectGraphMapping getFieldObjectGraphMapping() {
    return fieldObjectGraphMapping;
  }

  public Declaration getRelation() {
    return relation;
  }
}
