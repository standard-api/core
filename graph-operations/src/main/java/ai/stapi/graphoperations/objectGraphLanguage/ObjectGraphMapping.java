package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.declaration.Declaration;

public interface ObjectGraphMapping extends Declaration {

  String INTERFACE_UUID = "1eb45324-8162-44c5-ae87-104318446aed";

  GraphDescription getGraphDescription();
}
