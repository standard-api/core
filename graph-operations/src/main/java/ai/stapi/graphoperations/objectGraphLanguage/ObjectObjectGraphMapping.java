package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import java.util.Map;

public class ObjectObjectGraphMapping extends AbstractObjectGraphMapping {

  public static final String SERIALIZATION_TYPE = "ObjectObjectGraphMapping";
  private Map<String, ObjectFieldDefinition> fields;

  private ObjectObjectGraphMapping() {

  }

  public ObjectObjectGraphMapping(
      GraphDescription graphDescription,
      Map<String, ObjectFieldDefinition> fields
  ) {
    super(
        graphDescription,
        SERIALIZATION_TYPE
    );
    this.fields = fields;
  }

  public ObjectObjectGraphMapping(
      GraphDescription graphDescription,
      Map<String, ObjectFieldDefinition> fields,
      String serializationType
  ) {
    super(
        graphDescription,
        serializationType
    );
    this.fields = fields;
  }

  public Map<String, ObjectFieldDefinition> getFields() {
    return fields;
  }

}
