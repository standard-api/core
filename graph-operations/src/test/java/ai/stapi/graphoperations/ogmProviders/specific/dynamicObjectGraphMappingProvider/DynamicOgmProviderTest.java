package ai.stapi.graphoperations.ogmProviders.specific.dynamicObjectGraphMappingProvider;

import ai.stapi.graphoperations.fixtures.testsystem.TestSystemModelDefinitionsLoader;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectObjectGraphMapping;
import ai.stapi.graphoperations.ogmProviders.specific.dynamicObjectGraphMappingProvider.model.DynamicOgmProviderTestLoader;
import ai.stapi.schema.structureSchema.FieldDefinition;
import ai.stapi.schema.structureSchema.FieldType;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope({
    DynamicOgmProviderTestLoader.SCOPE,
    TestSystemModelDefinitionsLoader.SCOPE
})
class DynamicOgmProviderTest extends SchemaIntegrationTestCase {

  @Autowired
  private DynamicOgmProvider dynamicOgmProvider;

  @Test
  void itThrowsWhenTryingToGenerateOGMForNonExistingType() {
    Executable throwable = () -> this.dynamicOgmProvider.provideGraphMapping(
        "non-existing-serialization-type"
    );
    this.thenExceptionMessageApproved(RuntimeException.class, throwable);
  }

  @Test
  void itThrowsWhenTryingToGenerateOGMForPrimitiveType() {
    Executable throwable = () -> this.dynamicOgmProvider.provideGraphMapping("string");
    this.thenExceptionMessageApproved(RuntimeException.class, throwable);
  }

  @Test
  void itCanGenerateOgmForBoxedString() {
    var actual = this.dynamicOgmProvider.provideOgmForPrimitive(
        "BoxedString",
        new FieldDefinition(
            "exampleAttribute",
            0,
            "1",
            "",
            List.of(FieldType.asPlainType("string")),
            "Irrelevant"
        )
    );
    this.thenObjectApproved(actual);
  }

  @Test
  void itCanGenerateOgmForPrimitiveType() {
    var actual = this.dynamicOgmProvider.provideOgmForPrimitive(
        "decimal",
        new FieldDefinition(
            "exampleDecimalAttribute",
            0,
            "1",
            "",
            List.of(FieldType.asPlainType("decimal")),
            "Irrelevant"
        )
    );
    this.thenObjectApproved(actual);
  }

  @Test
  void itCanGenerateOGMForElementDefinition() {
    var result = this.dynamicOgmProvider.provideGraphMapping("ElementDefinition");
    this.thenObjectApproved(result);
  }

  @Test
  void itCanGenerateStructureDefinition() {
    var result = this.dynamicOgmProvider.provideGraphMapping("StructureDefinition");
    this.thenObjectApproved(result);
  }

  @Test
  void itCanGenerateOgmForLeafField() {
    var result =
        (ObjectObjectGraphMapping) this.dynamicOgmProvider.provideGraphMapping("Parameters");
    this.thenObjectApproved(result.getFields().get("id"));
  }

  @Test
  void itCanGenerateReferenceOgmForFieldWithContentReference() {
    var result = (ObjectObjectGraphMapping) this.dynamicOgmProvider.provideGraphMapping(
        "PlanDefinitionAction"
    );
    this.thenObjectApproved(result.getFields().get("action"));
  }

  @Test
  void itCanGenerateInterfaceOgmForFieldWithUnionType() {
    var result = (ObjectObjectGraphMapping) this.dynamicOgmProvider.provideGraphMapping(
        "Extension"
    );
    this.thenObjectApproved(result.getFields().get("value"));
  }
}