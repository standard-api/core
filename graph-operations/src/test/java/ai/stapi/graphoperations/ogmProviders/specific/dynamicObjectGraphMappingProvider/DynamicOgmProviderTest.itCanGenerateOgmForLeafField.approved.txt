{
  "serializationType" : "ObjectFieldDefinition",
  "fieldObjectGraphMapping" : {
    "serializationType" : "LeafObjectGraphMapping",
    "graphDescription" : {
      "serializationType" : "UuidIdentityDescription",
      "traversingType" : "value",
      "childGraphDescriptions" : [ ],
      "parameters" : { }
    }
  },
  "relation" : {
    "serializationType" : "NullGraphDescription",
    "traversingType" : "null",
    "childGraphDescriptions" : [ ]
  }
}